core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala

```diff
@@ -0,0 +1,130 @@
+
+import scala.collection.mutable
+
+import org.apache.spark.scheduler._
+
+private[spark] case class BlockUIData(
+    blockId: BlockId,
+    storageLevel: StorageLevel,
+    memSize: Long,
+    diskSize: Long,
+    externalBlockStoreSize: Long,
+    locations: Set[String])
+
+private[spark] class BlockStatusListener extends SparkListener {
+
+  private val blockManagers = new mutable.HashMap[BlockManagerId, mutable.HashSet[BlockId]]
+  private val blocks = new mutable.HashMap[BlockId, BlockUIData]
+
+  override def onBlockUpdated(blockUpdated: SparkListenerBlockUpdated): Unit = {
+    val blockId = blockUpdated.blockUpdatedInfo.blockId
+    if (!blockId.isInstanceOf[StreamBlockId]) {
+      // Now we only monitor StreamBlocks
+      return
+    }
+    val blockManagerId = blockUpdated.blockUpdatedInfo.blockManagerId
+    val storageLevel = blockUpdated.blockUpdatedInfo.storageLevel
+    val memSize = blockUpdated.blockUpdatedInfo.memSize
+    val diskSize = blockUpdated.blockUpdatedInfo.diskSize
+    val externalBlockStoreSize = blockUpdated.blockUpdatedInfo.externalBlockStoreSize
+
+    synchronized {
+      // Drop the update info if the block manager is not registered
+      blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+        if (storageLevel.isValid) {
+          blocksInBlockManager.add(blockId)
+          val location = blockManagerId.hostPort
+          val newLocations =
+            blocks.get(blockId).map(_.locations).getOrElse(Set.empty) + location
+          val newStorageLevel = StorageLevel(
+            useDisk = diskSize > 0,
+            useMemory = memSize > 0,
+            useOffHeap = externalBlockStoreSize > 0,
+            deserialized = storageLevel.deserialized,
+            replication = newLocations.size
+          )
+          blocks.put(blockId,
+            BlockUIData(
+              blockId,
+              newStorageLevel,
+              memSize,
+              diskSize,
+              externalBlockStoreSize,
+              newLocations))
+        } else {
+          // If isValid is not true, it means we should drop the block.
+          blocksInBlockManager -= blockId
+          removeBlockFromBlockManager(blockId, blockManagerId)
+        }
+      }
+    }
+  }
+
+  override def onBlockManagerAdded(blockManagerAdded: SparkListenerBlockManagerAdded): Unit = {
+    synchronized {
+      blockManagers.put(blockManagerAdded.blockManagerId, mutable.HashSet())
+    }
+  }
+
+  override def onBlockManagerRemoved(
+      blockManagerRemoved: SparkListenerBlockManagerRemoved): Unit = {
+    val blockManagerId = blockManagerRemoved.blockManagerId
+    synchronized {
+      blockManagers.remove(blockManagerId).foreach { blockIds =>
+        for (blockId <- blockIds) {
+          removeBlockFromBlockManager(blockId, blockManagerId)
+        }
+      }
+    }
+  }
+
+  private def removeBlockFromBlockManager(
+      blockId: BlockId, blockManagerId: BlockManagerId): Unit = {
```

Comment:
I think it is worth putting this method inside synchronized so that someone in future does not accidentally call this method without synchronizing.